src/iceberg/util/projection_util_internal.h

```diff
@@ -0,0 +1,640 @@
+          default:
+            return GenericTransform(std::move(ref), predicate, func);
+        }
+        std::unreachable();
+      }
+      default:
+        return NotSupported("{} is not a valid input type for temporal transform",
+                            func->source_type()->ToString());
+    }
+  }
+
+  static Result<std::unique_ptr<UnboundPredicate>> TruncateDecimal(
+      std::string_view name, const std::shared_ptr<BoundLiteralPredicate>& predicate,
+      const std::shared_ptr<TransformFunction>& func) {
+    const Literal& boundary = predicate->literal();
+    ICEBERG_ASSIGN_OR_RAISE(auto ref, NamedReference::Make(std::string(name)));
+
+    // For boundary adjustments, extract type info once
+    auto make_adjusted_literal = [&boundary](int adjustment) {
+      const auto& type = internal::checked_pointer_cast<DecimalType>(boundary.type());
+      Decimal adjusted = std::get<Decimal>(boundary.value()) + Decimal(adjustment);
+      return Literal::Decimal(adjusted.value(), type->precision(), type->scale());
+    };
+
+    switch (predicate->op()) {
+      case Expression::Operation::kLt: {
+        // adjust closed and then transform ltEq
+        ICEBERG_ASSIGN_OR_RAISE(auto transformed,
+                                func->Transform(make_adjusted_literal(-1)));
+        ICEBERG_ASSIGN_OR_RAISE(auto pred, UnboundPredicateImpl<BoundReference>::Make(
+                                               Expression::Operation::kLtEq,
+                                               std::move(ref), std::move(transformed)));
+        return pred;
+      }
+      case Expression::Operation::kGt: {
+        // adjust closed and then transform gtEq
+        ICEBERG_ASSIGN_OR_RAISE(auto transformed,
+                                func->Transform(make_adjusted_literal(1)));
+        ICEBERG_ASSIGN_OR_RAISE(auto pred, UnboundPredicateImpl<BoundReference>::Make(
+                                               Expression::Operation::kGtEq,
+                                               std::move(ref), std::move(transformed)));
+        return pred;
+      }
+      default:
+        return GenericTransform(std::move(ref), predicate, func);
+    }
+  }
+
+  static Result<std::unique_ptr<UnboundPredicate>> TruncateStringLiteral(
+      std::string_view name, const std::shared_ptr<BoundLiteralPredicate>& predicate,
+      const std::shared_ptr<TransformFunction>& func) {
+    const auto op = predicate->op();
+    if (op != Expression::Operation::kStartsWith &&
+        op != Expression::Operation::kNotStartsWith) {
+      return TruncateByteArray(name, predicate, func);
+    }
+
+    const auto& truncate_transform =
+        internal::checked_pointer_cast<TruncateTransform>(func);
+    const auto& str_value = std::get<std::string>(predicate->literal().value());
+    const auto width = truncate_transform->width();
+    ICEBERG_ASSIGN_OR_RAISE(auto ref, NamedReference::Make(std::string(name)));
+
+    if (str_value.length() < width) {
```

Comment:
Is this correct? `str_value.length()` is the number of bytes not the number of UTF-8 chars but `width` is the number of UTF-8 chars.

Reply:
Good catch, we should consider UTF-8, will fix.